kernel/src/vm/vmar/per_cpu_counter.rs

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! A fast and scalable per-CPU counter.
+
+use core::sync::atomic::{AtomicIsize, Ordering};
+
+use osdk_heap_allocator::{alloc_cpu_local, CpuLocalBox};
+use ostd::cpu::{all_cpus, CpuId};
+
+/// A fast, SMP-friendly, dynamically allocated, per-CPU counter.
+///
+/// Updating it is fast and scalable, but reading is slow and inaccurate.
+///
+// TODO: Reuse the code from [`osdk_frame_allocator::fast_smp_counter`],
+// which may need to extract that code into a separate crate that needs
+// to be published. Do that after we somehow stabilize the per-CPU counter.
+pub struct PerCpuCounter {
```

Comment:
You probably want to move this to `kernel/src/util` as this type is most likely to be useful to more modules other than `vmar`.